extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,165 @@
+
+/** @const {number} The number of milliseconds between checks to see if the response has arrived */
+const AD_BATCH_TICK_LENGTH = 50;
+
+/** @const {number} The number of ticks to wait for a response from the ad server before timing out */
+const AD_BATCH_TIMEOUT = 100;
+
+/** @const {!string} Tag name for ad batch manager implementation. */
+export const TAG_AD_BATCH_MANAGER = 'amp-ad-batch-manager';
+
+/**
+  * Get a batch manager that will be used to batch together elements with the same data-url
+  * Note a side effect: the first ad with a given data-url to be processed will have its isBatchMaster_ variable set to true
+  * @param {Element} e The element which is to be batched
+  * @param {string} url The URL identifying which batch
+  * @param {string} selector The selector to choose the things to be batched (e.g. 'amp-ad[type=imagead]')
+  * @returns {AmpAdBatchManager}
+  */
+export function getBatchManager(e, url, selector) {
+  // If this is our first imagead, create a map of responses for each value of 'data-url'
+  // This allows ads from multiple ad servers on the same page
+  if (!(e.win.hasOwnProperty('imageadBatchManagers'))) {
+    // Create an array of batch mangers. There will be one for each URL.
+    e.win.imageadBatchManagers = {};
+    // Scan the page to get a list of all image ads on this page, indexed by url.
+    // For performance reasons, we only do the scan once, the first time that a batch manager is constructed.
+    e.win.imageadElements = {};
+    const elements = document.querySelectorAll(selector);
+    for (let i = 0; i < elements.length; i++) {
+      const el = elements[i];
+      const url = el.getAttribute('data-url');
+      if (!(url in e.win.imageadElements)) {
+        e.win.imageadElements[url] = [];
+      }
+      e.win.imageadElements[url].push(el);
+    }
+  }
+  if (!(url in e.win.imageadBatchManagers)) {
+    // This is the first time we've seen this URL, so this will be the master for this batch
+    e.isBatchMaster_ = true;
+    e.win.imageadBatchManagers[url] = new AmpAdBatchManager(e);
+  } else {
+    e.isBatchMaster_ = false;
+  }
+  return e.win.imageadBatchManagers[url];
+}
+
+export class AmpAdBatchManager {
+
+  /** @param {!Element} element */
+  constructor(element) {
+
+    /** @private {string} The base URL of the ad server */
+    this.url_ = element.url_;
+
+    /** @private {string} The full URL of the ad server, including any slot ids */
+    this.fullurl_ = this.url_;
+
+    /** @private {string} The number of timer ticks remaining before we give up waiting for an ad server response */
+    this.timeout_ = AD_BATCH_TIMEOUT;
+
+    /** @public {Object} responseData Map of JSON responses indexed by slot */
+    this.responseData = null;
+
+    // Get a list of slots that are in use for this URL
+    const slots = [];
+    const elements = element.win.imageadElements[this.url_];
+    for (let i = 0; i < elements.length; i++) {
+      const e = elements[i];
```

Comment:
For sake of readability. 'e' can mean so many things. good name helps reader to remember where it is from